### DISCUSSION TITLE
Setting globally required `packageRules`, while allowing opt-in configuration

### DISCUSSION THREAD

**Discussion by jamietanna** [renovatebot/renovate#19166] [Request Help] [answered]
I'm looking to set up Renovate against a large organisation (~2100 repos), using a self-hosted setup.

I want to allow repositories to onboard their own configuration, to reduce the number of open PRs across the organisation and to prevent teams from overwhelm, with:

```javascript
// config.js
module.exports = {
  requireConfig: "required",
  onboarding: false,
};
```

However, I'd also like to set certain package managers to be set up in a way that allows us to allow us to ignore that, and always check certain types of dependency, for instance:

```javascript
// config.js
module.exports = {
  requireConfig: "required",
  onboarding: false,
  extends: [
    "{.....}:always"
  ]
};
```

Where `always.json` looks something like:

```json
{
  "requireConfig": "ignored",
  "packageRules": [
    {
      "matchManagers": [
        "docker-compose", "dockerfile", "kubernetes"
      ],
      "extends": [
        "schedule:monthly"
      ]
    }
  ]
}
```

Would this be the right way around doing this? Is there a better way to do this, and would this lead to any issues if a repo contains a `renovate.json`?

**Comment by rarkins** [comment]
You want:
1. To default to no repos running
2. To allow individual repos to opt-in by pushing a config
3. To bypass the above rule (1) for certain managers?

I'm not sure I under point (3) correctly, so please confirm or clarify

**Reply by jamietanna** [reply]
Yes that's correct - I'd like to make sure that all repos have certain managers enabled, and then if repos want additional managers, they can opt-in with their own configuration

**Reply by rarkins** [reply]
In that case I don't see the point of the root `requireConfig: "required",` setting because it would always be overridden with requireConfig=optional?

And if so, this means you'd run on your ~2100 repos although for those which don't find those limited managers it would exist quite quickly and users wouldn't notice anything? Meanwhile those who do have those managers or push their own config would get PRs?

**Reply by jamietanna** [reply]
Gotcha, that sounds reasonable. So something like:

```json
{
  "requireConfig": "ignored",
  "packageRules": [
    {
      "matchManagers": [
        "docker-compose", "dockerfile", "kubernetes"
      ],
      "extends": [
        "schedule:monthly"
      ]
    },
    {
      "matchManagers": [
        "*"
      ],
      "enabled": false
    }
  ]
}
```

And then for repos that are onboarded, they'd use a `default.json`:

```json
{
  "packageRules": [
    {
      "matchManagers": [
        "*"
      ],
      "enabled": true
    }
  ]
}
```

To make sure that disabled managers are re-enabled?

**Reply by rarkins** [reply]
You might be better with `enabledManagers` because it will be faster. If you use packageRules like you suggest, Renovate will extract _every_ manager, and then disable most. 

Also I suggest you put these defaults inside a company preset like you originally thought, so that repos which onboard themselves can just add it to `ignorePresets` to remove all the settings in one go.

**Reply by jamietanna** [reply]
Ah cool, so there'd be a global configuration such as (or in the `config.js`):

```json
{
  "requireConfig": "ignored",
  "extends": [
    "schedule:monthly"
  ],
  "enabledManagers": [
    "docker-compose", "dockerfile", "kubernetes"
  ]
}
```

Which makes sure that we're always updating those Docker dependencies.

And then we'd have a company-level preset, `default.json`:

```json
{
  "ignorePresets": true
}
```

If I wanted to set an `onboarding: false`, would that be able to be in the `config.js`, or would `ignorePresets` override it, so we'd need to move it to `default.json`?

**Reply by rarkins** [reply]
No, what I mean is that you set up a default company preset like `local>some-org/renovate-defaults` like:

```json
{
  "extends": [
    "schedule:monthly"
  ],
  "enabledManagers": [
    "docker-compose", "dockerfile", "kubernetes"
  ]
}
```

And your config.js is like:

```js
{
	"extends": ["local>some-org/renovate-defaults"],
    "requireConfig": "optional",
    "onboarding": false
}
```

Then in a repo their `renovate.json` can be like:

```json
{
	"ignorePresets": ["local>some-org/renovate-defaults"],
    "other config": "goes here"
}
```

i.e. the preset is stuff which active repos probably want to "reset" using the ignorePresets.

You can't have requireConfig=ignored or else it won't allow repos to define their own config

**Reply by jamietanna** [reply]
Thanks for this Rhys! Looks like it's mostly worked but `ignorePresets` doesn't seem to be getting used 🤔

So to confirm, I've got:

```javascript
{
module.exports = {
  requireConfig: "optional",
  onboarding: false,
  extends: [
    "github>some-org/renovate-config-presets:not-onboarded",
  ],
}
```

Then the `not-onboarded.json`:

```json
{
  "extends": [
    "github>some-org/renovate-config-presets:base"
  ],
  "dependencyDashboardApproval": true,
  "enabledManagers": [
    "regex"
  ]
}
```

Which pulls from `base.json` (which has org-specific settings):

```json
{}
```

Then I want to have a repo's `renovate.json` like so:

```json
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "extends": [
    "github>some-org/renovate-config-presets"
  ]
}
```

Which pulls the `default.json`:

```json
{
  "extends": [
    "config:base",
    "github>some-org/renovate-config-presets:base"
  ],
  "ignorePresets": [
    "github>some-org/renovate-config-presets:not-onboarded"
  ]
}
```

_However_, it looks like the `ignorePresets` doesn't seem to be taken into effect :thinking:

**Reply by jamietanna** [reply]
Could it be that repos that are onboarded with a `renovate.json` also need to set `ignorePresets` in their `renovate.json`, rather than hoping it'll be inherited? I guess so as it doesn't appear to be marked [in the docs](https://docs.renovatebot.com/configuration-options/#ignorepresets) as `mergeable: true`

**Reply by rarkins** [reply]
Please create a reproduction. If you're using Renovate O-P then this might be relevant too

**Reply by jamietanna** [reply]
Thanks, created https://github.com/jamietanna/renovate-config-19166-repo and https://github.com/jamietanna/renovate-config-19166 and just adding the [minimal repro docs](https://github.com/renovatebot/renovate/blob/main/docs/development/minimal-reproductions.md)

**Reply by jamietanna** [reply]
FYI I can reproduce this with Renovate OSS (`34.98.0`) as well as Renovate On-Prem (`3.0.0-rc.0`)

**Reply by rarkins** [reply]
I think it's related to the order in which the configs are resolved. In trace mode I see:

```
TRACE: Resolving preset "local>jamietanna/renovate-config-19166:not-onboarded" (repository=jamietanna/renovate-config-19166-repo)
TRACE: Resolving preset "github>jamietanna/renovate-config-19166:base" (repository=jamietanna/renovate-config-19166-repo)
TRACE: Resolving preset "github>jamietanna/renovate-config-19166" (repository=jamietanna/renovate-config-19166-repo)
TRACE: Resolving preset "config:base" (repository=jamietanna/renovate-config-19166-repo)
```

Because `not-onboarded` is supplied by the bot global config, the `extends` array in the repo consists of the not-onboarded config first, and then the default one. So by the time the ignorePresets is parsed from the default preset, it's "too late".

**Reply by jamietanna** [reply, marked as answer]
Ah interesting, so it looks like the `renovate.json` needs an `ignorePresets`, and then it works thanks!
